systemd-cryptenroll: add string aliases for tpm2 PCRs

1awesomeJ opened the pull request:
Fixes #26697. RFE.

1awesomeJ commented:
Reopens #27065 which I closed by accident.

Adding the integration test cases is causing conflicts with the main branch

could I do that in a separate PR?

1awesomeJ commented:
@bluca I could also push regardless sir. It's just four lines of test scripts. Perhaps you could help resolve the conflict sir.

bluca commented:
This doesn't look right, the changes in the test were already merged in main, you need to rebase

1awesomeJ commented:
> This doesn't look right, the changes in the test were already merged in main, you need to rebase

I was working across separate branches sir.
With two PR opened simultaneously


When I tried to sync this branch so I could pull changes, it created another commit on my PR.
A commit named "Merge..."

I tried resetting to an earlier commit,
I ended up closing  the PR by accident.

What looks feasible right now is for me to make all my changes on the other branch and open a new PR.

Except.of course, I could still learn from your review sir.

1awesomeJ commented:
Looks like I've been able to successfully rebase and add the integration tests.
Thanks for always insisting on the best sir.
In just about 3 weeks, My C programming, Linux, and git skills have been significantly improved.

1awesomeJ commented:
@bluca, @DaanDeMeyer , @keszybz , @mrc0mmand , @yuwata , @YHNdnzj .

Are there any modifications you would like me to make on this PR?

Kindly help review.

Thank you sirs.

poettering commented on src/basic/parse-util.c:
```diff
@@ -732,3 +733,23 @@ int parse_loadavg_fixed_point(const char *s, loadavg_t *ret) {
+        [PCR_SYSTEMD_BT]  = "systemdboot",
+        [PCR_SYSTEMD_STB] = "systemdstub",
+        [PCR_SHIM_CT]     = "shimcerts",
+        [PCR_LUKS_KEY]    = "lukskey",
```
please do not abbreviate unnecessarily.

please make sure that that the right-hand-side matches the name of the left-hand-side, except for maybe a common prefix. i.e. if the string is called "foo" or "bar" then the enum should be called XYZ_FOO or XYZ_BAR, but certainly not XYZ_F or XYZ_B or so.

Also, we usually use "-" as separator for enum identifiers.

1awesomeJ replied:
> please do not abbreviate unnecessarily.
> 
> please make sure that that the right-hand-side matches the name of the left-hand-side, except for maybe a common prefix. i.e. if the string is called "foo" or "bar" then the enum should be called XYZ_FOO or XYZ_BAR, but certainly not XYZ_F or XYZ_B or so.
> 
> Also, we usually use "-" as separator for enum identifiers.

Thank you sir. Sorry I'm not clear on the "-" separator. Do you mind showing an example? Thank you.

1awesomeJ commented on src/basic/parse-util.h:
```diff
@@ -11,6 +11,26 @@
+        PCR_SYSTEMD_BT,
+        PCR_SYSTEMD_STB,
+        PCR_SHIM_CT,
+        PCR_LUKS_KEY,
```
> this should not use automatic number assignment, since the numbers have meaning here. please specify the literal value for each entry.

Thank you sir. On it.

1awesomeJ replied:
> number 15 contains substantially more than the LUKS key...

Sorry to bother sir? but could you suggest a name for PCR 15? volumekey?
Thank you for taking the time.

I love reviews from this community.

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -251,55 +256,65 @@
 
               <row>
                 <entry>7</entry>
+                <entry>secureboot</entry>
```
"secure-boot-policy" maybe

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -251,55 +256,65 @@
 
               <row>
                 <entry>5</entry>
+                <entry>gptpartition</entry>
```
maybe "boot-gpt"

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -308,37 +323,44 @@
 
               <row>
                 <entry>10</entry>
+                <entry>imastate</entry>
```
just "ima"

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -308,37 +323,44 @@
 
               <row>
                 <entry>9</entry>
+                <entry>linuxkernel</entry>
```
"kernel-initrd"

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -308,37 +323,44 @@
 
               <row>
                 <entry>11</entry>
+                <entry>systemdkernel</entry>
```
"kernel-image"

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -308,37 +323,44 @@
 
               <row>
                 <entry>13</entry>
+                <entry>systemdstub</entry>
```
"sysext-image"

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -308,37 +323,44 @@
 
               <row>
                 <entry>14</entry>
+                <entry>shimcerts</entry>
```
"shim-policy"

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -308,37 +323,44 @@
 
               <row>
                 <entry>15</entry>
+                <entry>lukskey</entry>
```
"system-identity"

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -251,55 +256,65 @@
             <tbody>
               <row>
                 <entry>0</entry>
+                 <entry>corefirmware</entry>
```
"core-firmware"

1awesomeJ replied:
> "core-firmware"

This is an awesome awesome community. Thank you!

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -251,55 +256,65 @@
 
               <row>
                 <entry>4</entry>
+                <entry>bootloader</entry>
```
boot-loader

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -251,55 +256,65 @@
 
               <row>
                 <entry>6</entry>
+                <entry>powerstate</entry>
```
"power-state"

1awesomeJ commented:
What can I say @poettering, You're amazing. Thank you sir.

medhefgo commented on man/systemd-cryptenroll.xml:
```diff
@@ -251,55 +256,65 @@
 
               <row>
                 <entry>6</entry>
+                <entry>power-state</entry>
```
The spec says PCR6 is for `Host Platform Manufacturer Specific Measurements`. Assuming this is for power state events seems wrong to me.

1awesomeJ replied:
> The spec says PCR6 is for `Host Platform Manufacturer Specific Measurements`. Assuming this is for power state events seems wrong to me.

alright sir. "host-platform" 

Is that a good one sir?

1awesomeJ commented on man/systemd-cryptenroll.xml:
```diff
@@ -308,37 +323,44 @@
 
               <row>
                 <entry>12</entry>
+                <entry>kernel-configuration</entry>
```
> Either `platform-config` should not be shortened or this one should.

I would go with "platform-config" and "kernel-config" if you left the choice to me sir

1awesomeJ replied:
so users are a little less prone to typos?

1awesomeJ commented on man/systemd-cryptenroll.xml:
```diff
@@ -251,55 +256,65 @@
               </row>
 
               <row>
                 <entry>5</entry>
+                <entry>boot-gpt</entry>
```
> Should really be `boot-loader-config/boot-loader-data`.

Changing it accordingly

1awesomeJ commented:
@poettering , @medhefgo , I made the changes, kindly review.
Thanks

mrc0mmand commented on man/systemd-cryptenroll.xml:
```diff
@@ -394,7 +416,9 @@
-        be possible without losing access to all previously encrypted LUKS2 volumes.</para>
+        be possible without losing access to all previously encrypted LUKS2 volumes.
+        Like the <option>--tpm2-pcrs=</option>, string aliases as defined in the table above can also be used
+        to specify the registers. for instance <option>--tpm2-public-key-pcrs=boot-loader+system-identity</option></para>
```
Small nit:

```suggestion
        to specify the registers, for instance <option>--tpm2-public-key-pcrs=boot-loader+system-identity</option></para>
```

1awesomeJ replied:
> Small nit:

Good evening sir.
Please I'm not clear on what "Small nit" means. I can only tell you want me to make some changes. Do you mind elaborating a little more? Thank you

1awesomeJ replied:
Oh... I see it was a punctuation you helped correct. '.' , changed to the appropriate ','

Thank you so much sir.

mrc0mmand replied:
> > Small nit:
> 
> Good evening sir. Please I'm not clear on what "Small nit" means. I can only tell you want me to make some changes. Do you mind elaborating a little more? Thank you

The period is not correct here - either it should be a comma or the next word needs to be capitalized.

1awesomeJ replied:
I see it's a suggested changed.

1awesomeJ replied:
> Small nit:

Fixed sir.

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -251,55 +256,65 @@
 
               <row>
                 <entry>6</entry>
+                <entry>host-platform</entry>
```
hmm?

1awesomeJ replied:
> hmm?

What would you suggest sir, considering @medhefgo take on it?

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -251,55 +256,65 @@
 
               <row>
                 <entry>5</entry>
+                <entry>boot-loader-config</entry>
```
hmm, this isnt right?

1awesomeJ replied:
> hmm, this isnt right?

"boot-gpt" you suggested, @medhefgo suggested "boot-loader-config"

How about we use "boot-gpt-config"?

medhefgo replied:
The spec says it's boot loader confg/data. The fact that the gpt is measured into it is only part of it. So putting gpt into the name is not right.

poettering commented on src/basic/parse-util.h:
```diff
@@ -11,6 +11,26 @@
 
 typedef unsigned long loadavg_t;
 
+typedef enum Pcr {
```
maybe rename this PcrIndex

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -2354,14 +2354,20 @@ int tpm2_pcr_mask_from_string(const char *arg, uint32_t *ret_mask) {
-                        return log_error_errno(r, "Failed to parse PCR number: %s", pcr);
-                if (n >= TPM2_PCRS_MAX)
-                        return log_error_errno(SYNTHETIC_ERRNO(ERANGE),
+                if (pcr && ascii_isalpha(pcr[0])) {
```
please drop this, just use DEFINE_STRING_TABLE_LOOKUP_FROM_STRING_FALLBACK() if you use that instead of plain DEFINE_PRIVATE_STRING_TABLE_LOOKUP_FROM_STRING() then the numeric parsing will be done for you automatically

1awesomeJ replied:
> please drop this, just use DEFINE_STRING_TABLE_LOOKUP_FROM_STRING_FALLBACK() if you use that instead of plain DEFINE_PRIVATE_STRING_TABLE_LOOKUP_FROM_STRING() then the numeric parsing will be done for you automatically

Alright sir.
Will study it and implement

1awesomeJ replied:
> please drop this, just use DEFINE_STRING_TABLE_LOOKUP_FROM_STRING_FALLBACK() if you use that instead of plain DEFINE_PRIVATE_STRING_TABLE_LOOKUP_FROM_STRING() then the numeric parsing will be done for you automatically

@poettring I think I have implemented all the changes you requested apart from this.
Kindly confirm sir.

keszybz replied:
```C
DEFINE_STRING_TABLE_LOOKUP_WITH_FALLBACK(pcr_index, int, TPM2_PCRS_MAX);
```

poettering commented on src/basic/parse-util.h:
```diff
@@ -11,6 +11,26 @@
+        PCR_SYSEXT_IMAGE       = 13,
+        PCR_SHIM_POLICY        = 14,
+        PCR_SYSTEM_IDENTITY    = 15,
+        _PCR_MAX,
```
this should be set to TPM2_PCRS_MAX, because a TPM generally has 24 of these

1awesomeJ replied:
> this should be set to TPM2_PCRS_MAX, because a TPM generally has 24 of these

Alright sir.
Thank you.
If I may ask though when I set up the table pcr_table[_PCR_MAX] will it matter or not that I declare an array of 24 elements and only assign 15?
Totally asking to know, cos as you may have noticed, I'm only a prospective intern, getting a learning opportunity from the very top people in the profession.

Thank you sir.

1awesomeJ replied:
Finally, could I push the fixes a bit later, as I'm currently unable to.
Thank you.

keszybz commented on man/systemd-cryptenroll.xml:
```diff
@@ -237,7 +237,12 @@
         numeric PCR indexes in the range 0…23. If not used, defaults to PCR 7 only. If an empty string is
-        specified, binds the enrollment to no PCRs at all. PCRs allow binding the enrollment to specific
+        specified, binds the enrollment to no PCRs at all.
+        Users may also specify the specific registers they want to bind the enrollment to using string aliases
```
"Users may also specify the specific registers they want to bind the enrollment to" → "Registers may also be specified"

The part starting with "for instance" should be a new sentence.

keszybz commented on man/systemd-cryptenroll.xml:
```diff
@@ -237,7 +237,12 @@
+        specified, binds the enrollment to no PCRs at all.
+        Users may also specify the specific registers they want to bind the enrollment to using string aliases
+        for instance <option>--tpm2-pcrs=boot-loader+platform-config+boot-loader-config</option> to bind to the registers
+        <option> 4, 1, and 5</option> respectively. Check the PCR definitions table below for a full list
```
Strike "respectively".

Strike "and the registers they refer to".

keszybz commented on man/systemd-cryptenroll.xml:
```diff
@@ -237,7 +237,12 @@
+        for instance <option>--tpm2-pcrs=boot-loader+platform-config+boot-loader-config</option> to bind to the registers
+        <option> 4, 1, and 5</option> respectively. Check the PCR definitions table below for a full list
+        of available string aliases and the registers they refer to.
+        PCRs allow binding the enrollment to specific
```
New paragaraph here.

keszybz commented on man/systemd-cryptenroll.xml:
```diff
@@ -394,7 +416,9 @@
-        be possible without losing access to all previously encrypted LUKS2 volumes.</para>
+        be possible without losing access to all previously encrypted LUKS2 volumes.
+        Like the <option>--tpm2-pcrs=</option>, string aliases as defined in the table above can also be used
+        to specify the registers, for instance <option>--tpm2-public-key-pcrs=boot-loader+system-identity</option></para>
```
Dot missing at the end.

keszybz commented on src/test/test-parse-util.c:
```diff
@@ -922,4 +922,23 @@ TEST(parse_loadavg_fixed_point) {
+        assert_se(pcr_number_from_string((char*)"sysext-image") == 13);
+        assert_se(pcr_number_from_string((char*)"shim-policy") == 14);
+        assert_se(pcr_number_from_string((char*)"system-identity") == 15);
+        assert_se(pcr_number_from_string((char*)"hello") == -EINVAL);
```
Can we have a test for this option with `+`? That is more intresting…
Also `7+host-platform` and `6+host-platform+7`, etc.

1awesomeJ replied:
> Can we have a test for this option with `+`? That is more intresting… Also `7+host-platform` and `6+host-platform+7`, etc.

I have now added these test cases in testsuite-70.sh sir.

keszybz reviewed (changes requested):
Please change "added" to "add" and remove the dot in the commit title.
And change "fixes" to "Fixes".

keszybz commented on src/test/test-parse-util.c:
```diff
@@ -922,4 +922,23 @@ TEST(parse_loadavg_fixed_point) {
 }
 
+TEST(pcr_from_string) {
+        assert_se(pcr_number_from_string((char*)"core-firmware") == 0);
```
Why the cast?

1awesomeJ replied:
> Why the cast?

I wonder what I was thinking. @bluca has guided well enough on that.

keszybz commented on man/systemd-cryptenroll.xml:
```diff
@@ -308,37 +323,44 @@
 
               <row>
                 <entry>11</entry>
+                <entry>kernel-image-phase</entry>
```
It's not "phases of the kernel image". That makes little sense.
It is actually used for two distinct purposes: measurement of kernel, measurement of boot milestones.

What about `kernel-boot`? (I don't think we need to be *exact* in a complicated case like this one. Just give a general hint what the PCR is used for.)

1awesomeJ replied:
> It's not "phases of the kernel image". That makes little sense. It is actually used for two distinct purposes: measurement of kernel, measurement of boot milestones.
> 
> What about `kernel-boot`? (I don't think we need to be _exact_ in a complicated case like this one. Just give a general hint what the PCR is used for.)

I'm very happy lots of us are having inputs on this. Thank you so much sir.

keszybz commented on man/systemd-cryptenroll.xml:
```diff
@@ -308,37 +323,44 @@
 
               <row>
                 <entry>13</entry>
+                <entry>sysext-image</entry>
```
Just `systexts` ?

1awesomeJ replied:
> Just `systexts` ?

Alright sir.

keszybz commented on man/systemd-cryptenroll.xml:
```diff
@@ -394,7 +416,9 @@
-        be possible without losing access to all previously encrypted LUKS2 volumes.</para>
+        be possible without losing access to all previously encrypted LUKS2 volumes.
+        Like the <option>--tpm2-pcrs=</option>, string aliases as defined in the table above can also be used
+        to specify the registers, for instance <option>--tpm2-public-key-pcrs=boot-loader+system-identity.</option></para>
```
The dot needs to be *after* `</option>`.

keszybz commented on src/basic/parse-util.h:
```diff
@@ -11,6 +11,28 @@
+        PCR_SYSEXT_IMAGE       = 13,
+        PCR_SHIM_POLICY        = 14,
+        PCR_SYSTEM_IDENTITY    = 15,
+        _PCR_MAX               = TPM2_PCRS_MAX,
```
Please call this `_PCR_INDEX_MAX_DEFINED` and `_PCR_INDEX_INVALID`.

1awesomeJ replied:
> Please call this `_PCR_INDEX_MAX_DEFINED` and `_PCR_INDEX_INVALID`.

Certainly.

keszybz commented on test/units/testsuite-70.sh:
```diff
@@ -268,6 +268,22 @@ systemd-cryptenroll --tpm2-public-key-pcrs=key $img_2 && { echo 'unexpected succ
 
 systemd-cryptenroll --tpm2-pcrs=key $img_2 && { echo 'unexpected success'; exit 1; }
 
+systemd-cryptenroll --tpm2-pcrs=44+8 $img_2 && { echo 'unexpected success'; exit 1; }
+
+systemd-cryptenroll --tpm2-pcrs=8 $img_2
+
+systemd-cryptenroll --tpm2-pcrs=hello $img_2 && { echo 'unexpected success'; exit 1; }
+
+systemd-cryptenroll --tpm2-pcrs=boot-loader+boot-loader-config $img_2
+
+systemd-cryptenroll --tpm2-pcrs=7+host-platform $img_2
+
+systemd-cryptenroll --tpm2-pcrs=8+boot-loader-config+11 $img_2
+
+systemd-cryptenroll --tpm2-pcrs=5+boot-loader-config+11 $img_2
+
+systemd-cryptenroll --tpm2-pcrs=5+5+11 $img_2
```
This is going to be at least somewhat slow. I don't think there's any gain from enrolling the image a dozen times. Please move the tests for those combinations to the unit tests in `test-parse-util`, and actually check that the parse result is correct.

1awesomeJ replied:
> This is going to be at least somewhat slow. I don't think there's any gain from enrolling the image a dozen times. Please move the tests for those combinations to the unit tests in `test-parse-util`, and actually check that the parse result is correct.

Alright sir. I'll add the unit tests for tpm2_pcr_mask_from_string() function.

keszybz replied:
That's fine. We don't need a custom message.

1awesomeJ replied:
Awesome. Thank you sir.

1awesomeJ commented:
@poettering , @keszybz 
I have made the requested chnges sirs.
Kindly review.

medhefgo commented on man/systemd-cryptenroll.xml:
```diff
@@ -251,55 +256,65 @@
 
               <row>
                 <entry>6</entry>
+                <entry>power-events</entry>
                 <entry>Power state events; changes on system suspend/sleep</entry>
```
Again, the spec says PCR6 is "Host Platform Manufacturer Specific Measurements". I don't see anything about power events being measured into this. So this is just wrong or a big assumption/generalization.

1awesomeJ replied:
> Again, the spec says PCR6 is "Host Platform Manufacturer Specific Measurements". I don't see anything about power events being measured into this. So this is just wrong or a big assumption/generalization.

Thank you so much sir.
It's super exciting, that top experts here are taking their time to comment on this PR. You guys make me feel great.
Thank you sir. 
Let me quickly mention everyone else so we have a consensus.

I appreciate the time you take to even look at this PR, I appreciate even more, the fact that you're guiding.
Thank you sir.

keszybz replied:
We've been going back and forth on this. I proposed "power-events" because various wikis say that "Resume from S4 and S5 Power State Events" is recorded there. This was actually true in the past: "TCG PC Client Implementation for Conventional BIOS 1.2 revision 1.00" says "Entities to be Measured:
1. Resuming from S5 or S4". But "TCG PC Client Platform Firmware Profile Specification revision 1.05" says "Previously defined measurements for PCR[6] from the PC Client Implementation Specification for Conventional
BIOS have been deprecated". So let's assume that we know nothing about this PCR. Let's go back to the "host-platform" name then.

poettering replied:
I think we should just not assign a name to this PCR. Simply drop it from the table in docs and in the enum.

1awesomeJ replied:
> I think we should just not assign a name to this PCR. Simply drop it from the table in docs and in the enum.

My thoughts too at some point. I was like sine PR[8] is missing anyways, we could just leave out PCR[6] if it causes so much back and forth.

1awesomeJ commented:
@bluca @keszybz @mrc0mmand @poettering @medhefgo 
Please sirs, what would be a good consnsus for PCR 6.
@poettering suggested "power-satate", @keszybz suggested "power-events" @medhefgo suggested "host-platform" approximately?

What would be a good consensus sirs?

1awesomeJ commented:
@keszybz I have implemented all the changes requested, kindly review.

Thank you.

keszybz commented on src/test/test-parse-util.c:
```diff
@@ -922,4 +923,51 @@ TEST(parse_loadavg_fixed_point) {
+uint32_t k = 1024;
+
+assert_se(!tpm2_pcr_mask_from_string("7+power-events", &k));
+assert_se(!tpm2_pcr_mask_from_string("8+boot-loader,11", &k));
```
Well, the test test should actually test that the return value make sense.

keszybz replied:
Also, the intiailization of `k` is not needed.

1awesomeJ replied:
> Well, the test test should actually test that the return value make sense.

The function returns 0 on success or -EINVAL if it fails sir

1awesomeJ replied:
> Also, the intiailization of `k` is not needed.

Alright sir.

1awesomeJ replied:
@medhefgo, @keszybz has agreed to "host-platform".
Kindly check his explanation on the fact that TCG PC Client Platform Firmware Profile Specification revision 1.05 says "Previously defined measurements for PCR[6] from the PC Client Implementation Specification for Conventional
BIOS have been deprecated"

We'll use "host-platform".

1awesomeJ commented on src/test/test-parse-util.c:
```diff
@@ -922,4 +923,51 @@ TEST(parse_loadavg_fixed_point) {
+        assert_se(pcr_index_from_string("44") == -EINVAL);
+}
+
+TEST(tpm2_pcr_mask_from_string) {
+uint32_t k = 1024;
+
+assert_se(!tpm2_pcr_mask_from_string("7+power-events", &k));
+assert_se(!tpm2_pcr_mask_from_string("8+boot-loader,11", &k));
+assert_se(tpm2_pcr_mask_from_string("6+boot-loader,44", &k) == -EINVAL);
+assert_se(!tpm2_pcr_mask_from_string("7,shim-policy,4", &k));
+assert_se(!tpm2_pcr_mask_from_string("systexts,shim-policy+kernel-boot", &k));
+assert_se(tpm2_pcr_mask_from_string("systexts,shim+kernel-boot", &k) == -EINVAL);
+}
```
> Indentation?

On it sir.

1awesomeJ commented on src/test/test-parse-util.c:
```diff
@@ -922,4 +923,51 @@ TEST(parse_loadavg_fixed_point) {
+        assert_se(pcr_index_from_string("system-identity") == 15);
+        assert_se(pcr_index_from_string("15") == 15);
+        assert_se(pcr_index_from_string("hello") == -EINVAL);
+        assert_se(pcr_index_from_string("8") == 8);
```
> Please also test a negative number and a number between 16 and `TPM2_PCRS_MAX`.

On it sir.

1awesomeJ commented on man/systemd-cryptenroll.xml:
```diff
@@ -237,7 +237,12 @@
+        specified, binds the enrollment to no PCRs at all.
+        Registers may also be specified using string aliases.</para>
+        <para>For instance <option>--tpm2-pcrs=boot-loader+platform-config+boot-loader-config</option> to bind to the registers
+        <option> 4, 1, and 5</option>. Check the PCR definitions table below for a full list
```
> Drop this `<option>` markup. (This would land the string "4, 1, and 5" in the index, and we don't want that.)

Alright sir.

1awesomeJ commented on man/systemd-cryptenroll.xml:
```diff
@@ -394,7 +416,9 @@
         public key can be provided. The latter is hence more useful in scenarios where software updates shell
-        be possible without losing access to all previously encrypted LUKS2 volumes.</para>
+        be possible without losing access to all previously encrypted LUKS2 volumes.
+        Like the <option>--tpm2-pcrs=</option>, string aliases as defined in the table above can also be used
```
> "Like with" ?

Awesome

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -308,37 +323,44 @@
 
               <row>
                 <entry>13</entry>
+                <entry>systexts</entry>
```
"sysexts", not "systexts"

1awesomeJ replied:
It was "sysexts" originally based on your suggestion, I changed it based on @keszybz suggestion.

@keszybz, kindly note that I'll take this back to "sysexts"
Thank you.

It's been a wonderful privilege with all of you sirs.

poettering commented on src/basic/parse-util.c:
```diff
@@ -732,3 +733,23 @@ int parse_loadavg_fixed_point(const char *s, loadavg_t *ret) {
+        [PCR_IMA]                = "ima",
+        [PCR_KERNEL_BOOT]        = "kernel-boot",
+        [PCR_KERNEL_CONFIG]      = "kernel-config",
+        [PCR_SYSTEXTS]           = "systexts",
```
sysexts, not systexts

1awesomeJ replied:
> sysexts, not systexts

On it sir.

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -251,55 +256,65 @@
 
               <row>
                 <entry>3</entry>
+                <entry>external-firmware</entry>
```
external-configuration actually. The spec says:

> 3.3.4.4 PCR[3] – EFI Driver/Application Configuration
> This section contains the PCR[3] measurement requirements for a UEFI platform. In general, entities measure into PCR[3] data that is associated with code modules that are measured into PCR[2]. For example, on a UEFI platform, this includes but is not limited to UEFI variables defined and managed by drivers. See section 10.2.4 Measuring Industry Standard Tables and Data Structures. An example of information measured into PCR[3] is a SCSI controller’s configuration of its hard disks, e.g., RAID type, drive assignments, etc.

1awesomeJ replied:
> external-configuration actually. The spec says:
> 
> > 3.3.4.4 PCR[3] – EFI Driver/Application Configuration
> > This section contains the PCR[3] measurement requirements for a UEFI platform. In general, entities measure into PCR[3] data that is associated with code modules that are measured into PCR[2]. For example, on a UEFI platform, this includes but is not limited to UEFI variables defined and managed by drivers. See section 10.2.4 Measuring Industry Standard Tables and Data Structures. An example of information measured into PCR[3] is a SCSI controller’s configuration of its hard disks, e.g., RAID type, drive assignments, etc.

There's been no debate on PCR[3], the current option was totally from me. 
So it's switching swiftly to "external-configuration"

Thank you so much for the guidance.

poettering commented:
pcr 16 should probably be called "debug" as per spec

1awesomeJ commented:
> pcr 16 should probably be called "debug" as per spec

Definitely leaving out PCR[6], are you saying i should include PCR[16] sir?

poettering commented on src/basic/parse-util.h:
```diff
@@ -11,6 +11,28 @@
 
+#define TPM2_PCRS_MAX 24U
+
+typedef enum PcrIndex {
```
could you add a comment here, something like this:

```
/* The following names for PCRs 0…7 are based on the names in the "TCG PC Client Specific Platform Firmware Profile Specification" (https://trustedcomputinggroup.org/resource/pc-client-specific-platform-firmware-profile-specification/) */

1awesomeJ replied:
> could you add a comment here, something like this:
> 
> ```
> /* The following names for PCRs 0…7 are based on the names in the "TCG PC Client Specific Platform Firmware Profile Specification" (https://trustedcomputinggroup.org/resource/pc-client-specific-platform-firmware-profile-specification/) */
> ```

You are awesome sir. Thank you

From what I initially opened as PR, to what this is now, after @bluca, first made it refined from its very crude form, to all the inputs you guys have now added, It's awesome. Thank you everyone.

poettering commented on src/basic/parse-util.h:
```diff
@@ -11,6 +11,28 @@
+        PCR_BOOT_LOADER_CONFIG = 5,
+        PCR_POWER_EVENTS       = 6,
+        PCR_SECURE_BOOT_POLICY = 7,
+        PCR_KERNEL_INITRD      = 9,
```
and then insert a comment before this line, something like this:

```
/* The following names for PCRs 9…15 are based on the "Linux TPM PCR Registry"
(https://uapi-group.org/specifications/specs/linux_tpm_pcr_registry/) */
```
or so?

1awesomeJ replied:
> and then insert a comment before this line, something like this:
> 
> ```
> /* The following names for PCRs 9…15 are based on the "Linux TPM PCR Registry"
> (https://uapi-group.org/specifications/specs/linux_tpm_pcr_registry/) */
> ```
> 
> or so?

certainly sir.

1awesomeJ commented:
@keszybz, @medhefgo @poettering, 
Poettering and I think we should just leave out PCR[6] since it's definition has changed, and is now platform dependent.

It's awesome to have your attention on this.

Thank you.

poettering commented on src/basic/parse-util.h:
```diff
@@ -11,6 +11,28 @@
+        PCR_SYSTEXTS           = 13,
+        PCR_SHIM_POLICY        = 14,
+        PCR_SYSTEM_IDENTITY    = 15,
+        _PCR_INDEX_MAX_DEFINED = TPM2_PCRS_MAX,
```
as mentioned add this here:

```c
/* As per "TCG PC Client Specific Platform Firmware Profile Specification" again, see above */
PCR_DEBUG = 16,
```

1awesomeJ replied:
> as mentioned add this here:
> 
> ```c
> /* As per "TCG PC Client Specific Platform Firmware Profile Specification" again, see above */
> PCR_DEBUG = 16,
> ```

Awesome.

medhefgo commented:
> pcr 16 should probably be called "debug" as per spec

If we're skipping PCR6, then PCR16 should definitely be skipped. It should really not be used.

Also, the list is missing PCR23 (Application Support).

1awesomeJ commented:
> > pcr 16 should probably be called "debug" as per spec
> 
> If we're skipping PCR6, then PCR16 should definitely be skipped. It should really not be used.
> 
> Also, the list is missing PCR23 (Application Support).

Okay sir, We should leave out PCR[6] and PCR[16], but include PCR[23] as "application-support"

@poettering, Is that fine by you sir?

1awesomeJ commented:
@keszybz I have implemented all the changes you requested sir.

1awesomeJ commented:
@poettering I have implemented all the changes you took your time to point out sir.
In trying to have all of us on the same page, I added PCR[23] = "application-support" rather than PCR[16] sir. 
Every other thing you highlighted has been done sir.

1awesomeJ commented:
@medhefgo I have added the changes you requested sir. I dropped PCR[16] and added PCR[23].
Kindly review.

poettering commented:
> > pcr 16 should probably be called "debug" as per spec
> 
> If we're skipping PCR6, then PCR16 should definitely be skipped. It should really not be used.

why not? it appears to me the purpose of that PCR are very well defined and never changed. Seems riskless to me to give it a name.

1awesomeJ commented:
> > > pcr 16 should probably be called "debug" as per spec
> > 
> > 
> > If we're skipping PCR6, then PCR16 should definitely be skipped. It should really not be used.
> 
> why not? it appears to me the purpose of that PCR are very well defined and never changed. Seems riskless to me to give it a name.
@medhefgo , 
I think we should just do PCR[16] and PCR[23]
as Poettering mentioned, it's riskless for us to add it.

We've decided to leave out PCR[6] on the basis of the fact that its definition got changed and it's now platform dependent.
PCR[16] doesn't have such change.

In light of this I'll be adding PCR[16]
And as you requested sir, I already added PCR[23]

Let's all agree on these and move on with the change we're implenting. 

Perhaps we can push an update for PCR[6] whenever the docs provide more clarity that helps us all have thesame alias suggestion for it.

I feel overwhelmed that some of the brightest minds in the community are spending their time to review and comment on this PR.

Thank you so much sirs.

1awesomeJ commented:
@poettering 
I have implemented all the changes you requested.
kindly review.

1awesomeJ commented:
@medhefgo 
I have implemented all the changes you requested, kindly review.

1awesomeJ commented:
@keszybz 

I have implemented all the changes you requested.
Kindly review.

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -251,55 +256,65 @@
             <tbody>
               <row>
                 <entry>0</entry>
+                 <entry>platform-code</entry>
```
indentaion off by one space

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -251,55 +256,65 @@
                 <entry>6</entry>
+                <entry> </entry>
                 <entry>Power state events; changes on system suspend/sleep</entry>
               </row>
```
please drop this whole entry, as requested before. The current specs indicate this has no clear definition anymore, hence we shouldn't give it a name I am sure, and not document it here either.

1awesomeJ replied:
> please drop this whole entry, as requested before. The current specs indicate this has no clear definition anymore, hence we shouldn't give it a name I am sure, and not document it here either.

Alright sir. I didn't want to mess with the existing docs, but now that you've said to take it out, It's certainly getting lost.

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -308,39 +323,58 @@
+                <entry>Debug</entry>
+              </row>
+
+                <row>
```
indentation off by one char

poettering commented on src/basic/parse-util.h:
```diff
@@ -11,6 +11,32 @@
+        PCR_APPLICATION_SUPPORT = 23,
+        _PCR_INDEX_MAX_DEFINED  = TPM2_PCRS_MAX,
+        _PCR_INDEX_INVALID      = -EINVAL,
+} PcrIndex;
```
hmm, this shouldn't be defined in parse-util.h, which contains pretty generic parsers.

stuff specific to tpm should really be in tpm2-util.h.

1awesomeJ replied:
> hmm, this shouldn't be defined in parse-util.h, which contains pretty generic parsers.
> 
> stuff specific to tpm should really be in tpm2-util.h.

Awesome. Thanks for the guidance, you guys help me get better always. Thank you!

poettering commented on src/basic/parse-util.h:
```diff
@@ -11,6 +11,32 @@
 
+#define TPM2_PCRS_MAX 24U
+typedef enum PcrIndex {
+/* The following names for PCRs 0…7 are based on the names in the "TCG PC Client Specific Platform Firmware Profile Specification" (https://trustedcomputinggroup.org/resource/pc-client-specific-platform-firmware-profile-specification/) */
```
please line break, as usual

poettering commented on src/basic/parse-util.h:
```diff
@@ -11,6 +11,32 @@
 
 typedef unsigned long loadavg_t;
 
+#define TPM2_PCRS_MAX 24U
```
tpm2-util.h already defines that, don't define that again

poettering commented on src/test/test-parse-util.c:
```diff
@@ -922,4 +923,57 @@ TEST(parse_loadavg_fixed_point) {
+        assert_se(pcr_index_from_string("23") == 23);
+        assert_se(pcr_index_from_string("hello") == -EINVAL);
+        assert_se(pcr_index_from_string("8") == 8);
+        assert_se(pcr_index_from_string("44") == -EINVAL);
```
please test a negative pcr index

poettering commented on src/test/test-parse-util.c:
```diff
@@ -922,4 +923,57 @@ TEST(parse_loadavg_fixed_point) {
+}
+
+TEST(tpm2_pcr_mask_from_string) {
+uint32_t k;
```
indentation

poettering commented on src/test/test-parse-util.c:
```diff
@@ -922,4 +923,57 @@ TEST(parse_loadavg_fixed_point) {
+TEST(tpm2_pcr_mask_from_string) {
+uint32_t k;
+
+        assert_se(tpm2_pcr_mask_from_string("7+application-support", &k) == 0);
```
please check that k also has the right value afterwards, i.e. `(UINT32_C(1) << 7) | (UINT32_C(1) << 23)`

(similar for the other lines)

1awesomeJ commented:
Yay!!!
Good to have you back @poettering
Hope you had a great break.

Let me start making the fixes.

1awesomeJ commented:
@poettering,
I have pushed the required fixes. 
for this: "please check that k also has the right value afterwards, i.e. (UINT32_C(1) << 7) | (UINT32_C(1) << 23)"
I hardcoded the values based on the existing unit tests I found in test-tpm2.c

Kindly review sir.


Your reviews have been super helpful. (Your = @bluca @keszybz @poettering @medhefgo @mrc0mmand )
Thank you so much sirs!!!

keszybz commented:
OK, this has been through enough rounds, let's merge!

1awesomeJ commented:
> OK, this has been through enough rounds, let's merge!

Welcome back sir.
Did you all go on thesame break or to thesame event?

@poettering is here for the first time in 6 days, same as you.

And your own PRs too have been open since last week.


Thank you so much for all the guidance.

Please find time to review other ones I may open in the future.

You'll be helping train the next generation sir.
